### ISSUE TITLE
Rocket Mini to GDSII failed at step 15

### ISSUE THREAD

**Issue by Ryabkomv** [The-OpenROAD-Project/OpenLane#1126]
### Description
I have successfully generated GDSII file with open lane for picorv32 with configuration using docker container.
However when I try to get GDSII for Rocket chip with Tiny configuration I got error at step 15. 

```
[STEP 15]
[INFO]: Running Global Routing Resizer Timing Optimizations...
[ERROR]: during executing openroad script /openlane/scripts/openroad/resizer_routing_timing.tcl
[ERROR]: Exit code: 1
[ERROR]: full log: designs/ExampleRocketSystem/runs/RUN_2022.06.03_12.54.37/logs/routing/15-resizer.log
[ERROR]: Last 10 lines:
met2       Vertical      6182445       4937214          20.14%
met3       Horizontal    4121630       3293440          20.09%
met4       Vertical      2472978       1650582          33.26%
met5       Horizontal     824326        410880          50.16%
---------------------------------------------------------------

[INFO GRT-0101] Running extra iterations to remove overflow.
[ERROR GRT-0169] Net _0432988_: Invalid index for position (2590950, 3605250). Net degree: 6.
Error: resizer_routing_timing.tcl, 53 GRT-0169
child process exited abnormally

```

Googling I came to the following config.tcl

```
set ::env(DESIGN_NAME) ExampleRocketSystem
set ::env(FP_CORE_UTIL) 30

set ::env(CELL_PAD) 2

set ::env(PL_BASIC_PLACEMENT) 0
set ::env(PL_TARGET_DENSITY) 0.3

set ::env(GLB_RT_ADJUSTMENT) 0.2

set ::env(PL_ROUTABILITY_DRIVEN) 1

# Change if needed
set ::env(VERILOG_FILES) [glob $::env(DESIGN_DIR)/src/*.v]

# Fill this
set ::env(CLOCK_PERIOD) "10.0"
set ::env(CLOCK_PORT) "clk"

set filename $::env(DESIGN_DIR)/$::env(PDK)_$::env(STD_CELL_LIBRARY)_config.tcl
if { [file exists $filename] == 1} {
        source $filename
}
```

### Environment
```
docker --version
Docker version 20.10.16, build aa7e414


Please run the following set of commands in the OpenLane folder:

python3 ./env.py issue-survey
Kernel: Linux v5.4.0-113-generic
Distribution: ubuntu 20.04
Python: v3.8.10 (OK)
Container Engine: UNKNOWN vUNKNOWN (UNSUPPORTED)
OpenLane Git Version: 2653e6e4d69c8a8270c494e2ce61cbfd151fae38
pip: INSTALLED
pip:venv: INSTALLED
---
PDK Version Verification Status: OK
---
Git Log (Last 3 Commits)

2653e6e 2022-05-25T19:27:29+02:00 Synthesis Parameters, More Report Metrics (#1105) - Arya Reais-Parsi -  (HEAD -> master, tag: 2022.05.26_01.40.55, origin/master, origin/HEAD)
4298bf2 2022-05-25T17:42:21+02:00 Remove `endcap_cpp` option from `tapcell` (#1085) - Anton Blanchard -  ()
601eede 2022-05-21T18:47:55+02:00 Support Multi-Arch Docker Builds (#1075) - Mohamed Gaber -  (tag: 2022.05.22_02.07.28)
```

### Reproduction Material
* Upload a tarball containing the relevant design.
the used command 

`./flow.tcl -design ExampleRocketSystem`

### Expected behavior

Successful GDSII generation

[src.tar.gz](https://github.com/The-OpenROAD-Project/OpenLane/files/8842505/src.tar.gz)

[openroad_issue_reproducible](https://github.com/Ryabkomv/openroad_issue_reproducible)

**Comment by vijayank88** [comment]
@Ryabkomv 
`openroad_issue_reproducible` attachment missing

**Comment by vijayank88** [comment]
`Container Engine: UNKNOWN vUNKNOWN (UNSUPPORTED)` Also update docker version to latest one

@Donn is possible to print `issue-survey` with existing container version followed by `UNKNOWN` message. So its easy to convey the message to user.

**Comment by Ryabkomv** [comment]
Added sources, and docker version

**Comment by vijayank88** [comment]
@Ryabkomv 
Test case attached is not valid format.
`designs/ExampleRocketSystem/runs/RUN_2022.06.03_12.54.37/openroad_issue_reproducible` exist or not?

**Comment by Ryabkomv** [comment]
> @Ryabkomv Test case attached is not valid format. `designs/ExampleRocketSystem/runs/RUN_2022.06.03_12.54.37/openroad_issue_reproducible` exist or not?
> 
> If not exist Please package a standalone test case with https://github.com/The-OpenROAD-Project/OpenLane/blob/master/docs/source/using_or_issue.md

Added corresponding folder as an archive at [github](https://github.com/Ryabkomv/openroad_issue_reproducible). It pretty lengthy.

**Comment by vijayank88** [comment]
@Ryabkomv 
can you share me git repo link, where you found this project?

**Comment by vijayank88** [comment]
If you're using default configuration try increase `FP_CORE_UTIL` to 30 and try

**Comment by Ryabkomv** [comment]
> @Ryabkomv can you share me git repo link, where you found this project?

https://github.com/chipsalliance/rocket-chip
I used TinyConfig configuration

**Comment by Ryabkomv** [comment]
> If you're using default configuration try increase `FP_CORE_UTIL` to 30 and try

Yes, I copied my config and  `FP_CORE_UTIL` is already set to 30

**Comment by vijayank88** [comment]
Can you share src/*.v files?

**Comment by Ryabkomv** [comment]
> Can you share src/*.v files?

These files are already attached to the post, can you download them?

**Comment by vijayank88** [comment]
`CLOCK_PORT` is `clock` right?

**Comment by Ryabkomv** [comment]
> set ::env(CLOCK_PORT) "clk"

Should be...
So you mean I need to change
`set ::env(CLOCK_PORT) "clk"`
for
`set ::env(CLOCK_PORT) "clock"`
right?

**Comment by vijayank88** [comment]
yes. I am also running at my end with changes. Will update you the status.

**Comment by Ryabkomv** [comment]
oh, looks like a stupid mistake...I will rerun the flow with updated config file.
By the way, Is there a way to run the flow step by step? Because the process is pretty time consuming, in case of failure at some stage I'm looking to make modifications in config.tcl and re-run just failed stage.

**Comment by vijayank88** [comment]
Use `./flow.tcl -interactive`

**Comment by vijayank88** [comment]
@maliberty  Can you please  look into this?

**Comment by vijayank88** [comment]
@Ryabkomv Attach latest run test case

**Comment by Ryabkomv** [comment]
Unfortunately got the issue at step 15.
updated test case at [git repository](https://github.com/Ryabkomv/openroad_issue_reproducible)

**Comment by maliberty** [comment]
I am out this week, perhaps @luis201420 can help

**Comment by vijayank88** [comment]
@luis201420 
do you had time to look into this issue?

**Comment by vijayank88** [comment]
@donn 
Is there any configuration available to change `set_routing_alpha 0` in openlane flow?

@luis201420 
what is default setting for `set_routing_alpha`? In which scenario we need to pass this variable?

**Comment by maliberty** [comment]
The alpha=0 is a workaround to a router problem that we are fixing.  You would need it when you hit this error.

**Comment by Ryabkomv** [comment]
Thanks for suggestion.
I wonder if I need to set variable in config file? What is there name of the variable? Should it looks like:
`set ::env(set_routing_alpha) 0 ?`
I cannot see such variable in [this list](https://github.com/The-OpenROAD-Project/OpenLane/blob/master/configuration/README.md)

**Comment by vijayank88** [comment]
@Ryabkomv 
Check for following file in your local path: https://github.com/The-OpenROAD-Project/OpenLane/blob/master/scripts/openroad/groute.tcl

In line 45 add following lines and run the flow.
`set ::env(set_routing_alpha) 0`

**Comment by Ryabkomv** [comment]
Thank, will do

**Comment by vijayank88** [comment]
@Ryabkomv 
updated command format

**Comment by Ryabkomv** [comment]
Got same error

[STEP 15]
[INFO]: Running Global Routing Resizer Timing Optimizations...
[ERROR]: during executing openroad script /openlane/scripts/openroad/resizer_routing_timing.tcl
[ERROR]: Exit code: 1
[ERROR]: full log: designs/ExampleRocketSystem/runs/RUN_2022.06.16_06.21.38/logs/routing/15-resizer.log
[ERROR]: Last 10 lines:
met2       Vertical      6182445       4528494          26.75%
met3       Horizontal    4121630       2882560          30.06%
met4       Vertical      2472978       1614742          34.70%
met5       Horizontal     824326        410880          50.16%


[INFO GRT-0101] Running extra iterations to remove overflow.
[ERROR GRT-0169] Net _0430322_: Invalid index for position (1983750, 3805350). Net degree: 6.
Error: resizer_routing_timing.tcl, 53 GRT-0169
child process exited abnormally


part of groute.tcl with set set_routing_alpha

```
source $::env(SCRIPTS_DIR)/openroad/set_routing_layers.tcl

set_macro_extension $::env(GLB_RT_MACRO_EXTENSION)
**set ::env(set_routing_alpha) 0**
source $::env(SCRIPTS_DIR)/openroad/layer_adjustments.tcl

set arg_list [list]
lappend arg_list -congestion_iterations $::env(GLB_RT_OVERFLOW_ITERS)
lappend arg_list -verbose
if { $::env(GLB_RT_ALLOW_CONGESTION) == 1 } {
    lappend arg_list -allow_congestion
}

```

**Comment by Ryabkomv** [comment]
oh, sorry

**Comment by Ryabkomv** [comment]
still failing at the same step with the same error

**Comment by vijayank88** [comment]
@luis201420 
If I use `set_routing_alpha 0` before `global_route` command openroad crashed.
```
[INFO GRT-0101] Running extra iterations to remove overflow.
[INFO GRT-0103] Extra Run for hard benchmark.
/opt/rh/devtoolset-8/root/usr/include/c++/8/bits/stl_vector.h:932: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](std::vector<_Tp, _Alloc>::size_type) [with _Tp = int; _Alloc = std::allocator<int>; std::vector<_Tp, _Alloc>::reference = int&; std::vector<_Tp, _Alloc>::size_type = long unsigned int]: Assertion '__builtin_expect(__n < this->size(), true)' failed.
Stack trace:
 0# 0x0000000000BF53D6 in /home/vijayan/OPENROAD_FLOW/OpenROAD_FLOW/OpenROAD/build/src/openroad
 1# 0x00007FD67ADD0400 in /lib64/libc.so.6
 2# gsignal in /lib64/libc.so.6
 3# abort in /lib64/libc.so.6
 4# boost::stacktrace::detail::this_thread_frames::collect(void const**, unsigned long, unsigned long) in /home/vijayan/OPENROAD_FLOW/OpenROAD_FLOW/OpenROAD/build/src/openroad
 5# 0x0000000002724FC2 in /home/vijayan/OPENROAD_FLOW/OpenROAD_FLOW/OpenROAD/build/src/openroad
 6# grt::FastRouteCore::updateRouteType1(int, grt::TreeNode const*, int, int, int, int, int, grt::TreeEdge*, int, int) in /home/vijayan/OPENROAD_FLOW/OpenROAD_FLOW/OpenROAD/build/src/openroad
 7# grt::FastRouteCore::mazeRouteMSMD(int, int, float, int, int, bool, int, float, int, int, int) in /home/vijayan/OPENROAD_FLOW/OpenROAD_FLOW/OpenROAD/build/src/openroad
 8# grt::FastRouteCore::run() in /home/vijayan/OPENROAD_FLOW/OpenROAD_FLOW/OpenROAD/build/src/openroad
 9# grt::GlobalRouter::findRouting(std::vector<grt::Net*, std::allocator<grt::Net*> >&, int, int) in /home/vijayan/OPENROAD_FLOW/OpenROAD_FLOW/OpenROAD/build/src/openroad
10# grt::GlobalRouter::globalRoute() in /home/vijayan/OPENROAD_FLOW/OpenROAD_FLOW/OpenROAD/build/src/openroad
11# 0x0000000001416B8F in /home/vijayan/OPENROAD_FLOW/OpenROAD_FLOW/OpenROAD/build/src/openroad
```

**Comment by luis201420** [comment]
@vijayank88 I cloned the [github repository](https://github.com/Ryabkomv/openroad_issue_reproducible).
I unzipped the file **openroad_issue_reproducible.tar.gz** and I ran openROAD on the directory "openroad_issue_reproducible/home/ryabko/OpenLane/designs/ExampleRocketSystem/runs/RUN_2022.06.07_10.44.44/openroad_issue_reproducible/":
`./openroad -no_init run.tcl`
And I got the error:
```
[ERROR GRT-0169] Net _0428215_: Invalid index for position (2680650, 2176950). Net degree: 5.
[ERROR GUI-0070] Error: resizer_routing_timing.tcl, 54 GRT-0169
```
Then, I added on **openlane/scripts/openroad/resizer_routing_timing.tcl** file before the global_route command (Line 53-54) the `set_routing_alpha 0` command to solve the invalid index error.
With this command the design finishes the extra overflow iterations without problems but in the end the output gives **Routing congestion too high**.

**Comment by dralabeing** [comment]
@luis201420  What is the next step here?

**Comment by dralabeing** [comment]
@vijayank88  create a new GH with the ORFS crash you see on this test case. Please re-test and If there is no congestion issues close this one.

**Comment by Ryabkomv** [comment]
Dear all, I'm little bit lost. Is there a way to resolve the issue? 
Is so, can you kindly let me know the needed config file for that?
In my case I still not able to pass stage 15. 
I tried default config for rocket chip and still have issue. The only successes was with picorv32 processor and simple components.

**Comment by vijayank88** [comment]
@Ryabkomv 
Do you have any idea how much clock frequency it can be operated?

**Comment by Ryabkomv** [comment]
This is good question. I used 10 MHz. I thought it is pretty reasonable. Never used less. 
Is it too fast?

**Comment by vijayank88** [comment]
I tried with 30ns = 33.33MHz. Its passing with ORFS flow.
OpenLane Its crashed during global routing. Need to debug that.

**Comment by vijayank88** [comment]
Final routing view from ORFS flow.
![mini_rocket_gui](https://user-images.githubusercontent.com/85787913/176108061-f4c94c7f-b94d-47f2-bc5d-015360f4c7cb.png)

**Comment by vijayank88** [comment]
https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts

**Comment by vijayank88** [comment]
> @vijayank88 create a new GH with the ORFS crash you see on this test case. Please re-test and If there is no congestion issues close this one.

@dralabeing With ORFS flow mini rocket chip routing got completed. I think that crash due to memory issue. took 25GB RAM for post routing timing check and report.

**Comment by Ryabkomv** [comment]
By the way,
in my case I have about 500Gb of RAM

**Comment by antonblanchard** [comment]
@Ryabkomv using either DFFRAM or OpenRAM for those arrays would likely help a lot.

**Comment by Ryabkomv** [comment]
Can you let me know which file you need?
 I think I provided the link [link](https://github.com/Ryabkomv/openroad_issue_reproducible) to the tarball with all possible logs along with config file containing all used parameters.

**Comment by maliberty** [comment]
@Ryabkomv **no** that tarball does not contain all log files.

**Comment by Ryabkomv** [comment]
Please, let me know which log file you need. Where are they located?
Sorry for my questions, but I am a newbie with OpenLane, just using step by
step instructions.

вт, 28 июн. 2022 г. в 22:44, Matt Liberty ***@***.***>:

**Comment by Ryabkomv** [comment]
Ok, I've cleaned all the runs history, so will make a fresh run and upload
in a few hours.

ср, 29 июн. 2022 г. в 10:22, vijayan ***@***.***>:

**Comment by Ryabkomv** [comment]
Please, use the [link ](https://github.com/Ryabkomv/openroad_issue_reproducible)to find all the file with log folder.

**Comment by Ryabkomv** [comment]
Guys, were the logs I uploaded enough? Any others files needed?

**Comment by vijayank88** [comment]
@Ryabkomv Have you checked your RTL to update memory blocks with OpenRAM/DFFRAM change?

**Comment by Ryabkomv** [comment]
No, I didn't. I thought rocket project can be synthetized as it is.
Is the a simple way to update to OpenRAM/DFFRAM.
Should it be done through config file?

**Comment by vijayank88** [comment]
Source RTL update can't do with config.tcl.

**Comment by Ryabkomv** [comment]
Ok,
so in summary, current version of OpenLane can't synthetize rocket core with the [source ](https://github.com/chipsalliance/rocket-chip) as it is, right?
And the workaround is to dive into RTL?

Just for information, my interest in RISC-V is not about designing come advanced cores, I want to use standard core and supplement it with photonic accelerator through ROCC interface. So I did not expect to modify any RTL, just work on GDSII files from RISC-V  and photonic chip to connect corresponding wires.

**Comment by vijayank88** [comment]
@Ryabkomv 
Please don't mislead it.
OpenLane synthesize the current rocket core. Its facing issue while running global routing.
So i am asking whether is it possible to optimize source RTL?

**Comment by Ryabkomv** [comment]
I'm sorry if I said something wrong.  I meant the process of getting GDSII from open source rocket core written in chisel.
I will check what is DFFRAM/OpenRAM and how to use it in the project. RTL is a huge verilog file, If I will be able to find step by step instruction I will do. Otherwise it seems impossible in reasonable time with my background.

**Comment by tspyrou** [comment]
@vijayank88 We should treat the RTL as fixed in this case. Could you try it with a lower utilization?

**Comment by maliberty** [comment]
Sorry I've been off on an important task.  I looked at 8-resizer.log and I see
```
[INFO ODB-0131]     Created 1098296 components and 6431219 component-terminals.
...
[INFO RSZ-0038] Inserted 4000 buffers in 3063 nets.
[INFO RSZ-0039] Resized 790199 instances.
```
The vast majority of your gates are needing to be upsized.  This is likely to be the cause of the excess congestion (unfortunately it doesn't record the area increase due to sizing).  

In 7-global.log you have a fair number of ERC violations.
```
max slew violation count 86353
max fanout violation count 2
max cap violation count 14436
```
Its odd that at the end of synthesis you have very few (730 max slew).  Something seem off that synthesis see such a different view.  Perhaps there is an issue with wire load models?

Also a fair amount of slack repair needed:
```
tns -664458.94
wns -22.90
```
after CTS I see:
```
max fanout violation count 19090
```
I'm not sure why the max fanout is set to 5 but it seems quite low.  It does seem to be an OL default rather than specific to your design.

In all it seems like the design requires a lot of repair and that is increasing the utilization.  Are you trying to hit a particularly high frequency?  The debugging really needs to look at the full flow.   Is there a repo with the full design collateral necessary to run OL?

**Comment by antonblanchard** [comment]
> I'm not sure why the max fanout is set to 5 but it seems quite low. It does seem to be an OL default rather than specific to your design.

I think 5 is overly pessimistic. I've been overriding the default, and using 10 (which still may be low).

**Comment by antonblanchard** [comment]
I created https://github.com/The-OpenROAD-Project/OpenLane/pull/1187

**Comment by vijayank88** [comment]
> @vijayank88 We should treat the RTL as fixed in this case. Could you try it with a lower utilization?

@tspyrou 
Yes. I am using RTL as it is and Re-Running the flow again with changes suggested above. Update you once got flow status.

**Comment by vijayank88** [comment]
@Ryabkomv 
Update openlane to latest commit. 
https://github.com/The-OpenROAD-Project/OpenLane/pull/1172/files merge this update into `scripts/openroad/replace.tcl`
Use following `config.tcl` will resolve congestion issue and started running DRT now.
```
# User config
set ::env(DESIGN_NAME) ExampleRocketSystem
set ::env(FP_SIZING) "absolute"
set ::env(DIE_AREA) {0.0 0.0 6937.49 6948.21}
set ::env(CORE_AREA) {5.52 10.88 6931.7 6935.04}
set ::env(FP_CORE_UTIL) 0.01
set ::env(DIODE_INSERTION_STRATEGY) {4}

set ::env(CELL_PAD) 1

#set ::env(PL_BASIC_PLACEMENT) 0
set ::env(PL_TARGET_DENSITY) 0.22

set ::env(GLB_RT_ADJUSTMENT) 0.15

#set ::env(PL_ROUTABILITY_DRIVEN) 1
# Change if needed
set ::env(VERILOG_FILES) [glob $::env(DESIGN_DIR)/src/*.v]
set ::env(GLB_RESIZER_TIMING_OPTIMIZATIONS) {0}
# Fill this
set ::env(CLOCK_PERIOD) "30.0"
set ::env(CLOCK_PORT) "clock"
set ::env(SYNTH_MAX_FANOUT) 10
set filename $::env(DESIGN_DIR)/$::env(PDK)_$::env(STD_CELL_LIBRARY)_config.tcl
if { [file exists $filename] == 1} {
	source $filename
```

**Comment by Ryabkomv** [comment]
Hello.
Now I have crush at the 7th stage.
I have updated [repo ](https://github.com/Ryabkomv/openroad_issue_reproducible)with my new replace.tcl and uploaded the log file of the corresponding stage.

**Comment by vijayank88** [comment]
@Ryabkomv 
can you try with above config.tcl that I attached?

**Comment by Ryabkomv** [comment]
@vijayank88 
I have updated OpenLane following the [instruction](https://github.com/The-OpenROAD-Project/OpenLane#updating-openlane).
When I run `make test` I got error with updated [replace.tcl](https://github.com/The-OpenROAD-Project/OpenLane/blob/0aa9439bab9a8e5081678f04aff760704d369780/scripts/openroad/replace.tcl). So I think current issue with this file. Can you confirm that the file is correct?
With the old one (before modification) all tests are passed.
And I do use the config.tcl you've sent.

**Comment by Ryabkomv** [comment]
I'm sorry, but after OpenLane update I've got new issue:
The version of openroad_app installed in the enviroment does not match the one required by OpenLane flow scripts.

Should I update the docker container which includes oenroad?

**Comment by vijayank88** [comment]
do you know how to update `openroad_app` commit right?

Try this way: `OPENLANE_IMAGE_NAME=efabless/openlane:current make mount`

**Comment by Ryabkomv** [comment]
In my case the crush at step 7 (Global placement)

**Comment by vijayank88** [comment]
@Ryabkomv 
As per your log still using `OpenROAD 79a46b62da64bbebc18f06b20c42211046de719a ` commit.

I suggested to use latest commit from [OpenROAD](https://github.com/The-OpenROAD-Project/OpenROAD):  `b3f2819780aa0a6e0000f0fd6f5b76fd951ceb52`

**Comment by Ryabkomv** [comment]
I have updated OpenLane yesterday. log file was not updated. Now I've [updated ](https://github.com/Ryabkomv/openroad_issue_reproducible/blob/master/7-global.log)it

**Comment by vijayank88** [comment]
Just debugged through teams meeting,  `config.tcl` settings over-written by `sky130A_sky130_fd_sc_hd_config.tcl` configuration  as its contains some default setting. @Ryabkomv removed those configuration and running the flow. Hope this time able to get final routing results.

**Comment by vijayank88** [comment]
@Ryabkomv 
Any updates on flow status?

**Comment by Ryabkomv** [comment]
It is on step #39 now. DRC check

**Comment by Ryabkomv** [comment]
Finally I got error on step 41 CVC/
`There are hold violations in the design at the typical corner`

**Comment by vijayank88** [comment]
@Ryabkomv 
Can you share `logs/signoff/xx-parasitics_sta.log`?

**Comment by Ryabkomv** [comment]
I have updated all [logs](https://github.com/Ryabkomv/openroad_issue_reproducible/blob/master/logs.tar.gz)

**Comment by Ryabkomv** [comment]
> I did not read the whole thread. But I noticed some abnormalities worth investigating:
> 
> 1. CLOCK_PORT is supposed to be "clock". Assuming RockerChip uses Chisel default clock naming. I burnt there a couple of times.
> 2. the density is set to the corresponding value as the FP CORE UTIL. This may cause congestion issues down the line. I would recommend increasing the density to be at least 10% higher than core util. In the design according to the logs, the density is much higher than it is set to.
> 3. Always create a new design using -init_design. Copy pasting some designs folder may cause the wrong files to be picked, which you did not intend to. Burnt there many times.
> 4. the RocketChip is supposed to have the RAM cells replaced with tech-specific. Otherwise, the area usage is going to be insane => long runtimes and a lot of memory usage. It generates the config file, which can be used to configure the DFFRAM.

@armleo thanks for your suggestions.
Can you kindly share the link with the example to use DFFRAM with Rocket Core?

**Comment by vijayank88** [comment]
@Ryabkomv 
As per guideline one issue to be discussed at a time.

As congestion issue got fixed and this issue should be closed. You can open a new issue if there is any other problem.

Regarding hold violation fix:
```
set ::env(GLB_RESIZER_HOLD_SLACK_MARGIN) {0.1}
```
Change the margin as per `wns` and re-run the flow to meet timing.
If no other issues mean, you may close this issue.

**Comment by Ryabkomv** [comment]
@vijayank88 you suggested to set fixed size of the chip. I wonder if it is appropriate way? For me, with no experience in VLSI, I hardly can imagine how to calculate the size before starting the flow.

**Comment by vijayank88** [comment]
```
set ::env(FP_SIZING) "absolute"
set ::env(DIE_AREA) {0.0 0.0 6937.49 6948.21}
set ::env(CORE_AREA) {5.52 10.88 6931.7 6935.04}
```
Removing above lines from `config.tcl` OpenLane flow uses relative size based die area calculation.

With our initial try tool calculated value is:
```

set ::env(DIE_AREA) {0.0 0.0 5937.49 5948.21}
set ::env(CORE_AREA) {5.52 10.88 5931.7 5935.04}
```

Now you can try to optimize design as per need

**Comment by Ryabkomv** [comment]
It looks I'm lost.
With new parameters the flow crushes at stage 6 because of low disk space. It uses more than 300 GB!!! I've only commented absolute sizing and added:
```
set ::env(PL_RESIZER_HOLD_SLACK_MARGIN) {0.1}
set ::env(GLB_RESIZER_HOLD_SLACK_MARGIN) {0.1}
```

**Comment by vijayank88** [comment]
@Ryabkomv 
Check `runs` directory. Keep only previous successful `RUNSxxx` and remove fail RUNS. So you will get free space.

**Comment by Ryabkomv** [comment]
Thanks, but I do remove all the runs. Only the last run consumes >300GB

**Comment by vijayank88** [comment]
@Ryabkomv 
Any updates?

**Comment by Ryabkomv** [comment]
Well, I'm still trying to succeed by changing config parameters, but I have no successful run completed yet!
  
I'm curious if it possible to add an example of successful application of OpenLane with some default parameters to Rocket core generated by chipyard?

I'm not sure if my case so specific, or only VLSI experts can use such tools? I mean try and failure adjusting parameters in config file.

By commenting absolute `FP_SIZING ` the flow fails with disk space error. RUN folder consumes more than 300 GB, can you suggest why it happens?

I returned back absolute sizing and started the flow with the only change 

```
set ::env(PL_RESIZER_HOLD_SLACK_MARGIN) {0.1}
set ::env(GLB_RESIZER_HOLD_SLACK_MARGIN) {0.1}
```

**Comment by vijayank88** [comment]
@Ryabkomv 
Goto `OpenLane` folder and do following to commit latest updates from OpenLane.
```
git checkout master
git pull
make
```
If you're running out of memory remove previous `runs` from your directory.
As per latest OpenLane use following `config.tcl` to run rocket chip
```
# User config
set ::env(DESIGN_NAME) ExampleRocketSystem
set ::env(FP_CORE_UTIL) 0.01
set ::env(DIODE_INSERTION_STRATEGY) {4}

set ::env(DPL_CELL_PADDING) 1


set ::env(PL_TARGET_DENSITY) 0.22

set ::env(GRT_ADJUSTMENT) 0.15

# Change if needed
set ::env(VERILOG_FILES) [glob $::env(DESIGN_DIR)/src/*.v]

# Fill this
set ::env(CLOCK_PERIOD) "30.0"
set ::env(CLOCK_PORT) "clock"
set ::env(SYNTH_MAX_FANOUT) 10

```

**Comment by Ryabkomv** [comment]
@vijayank88 thanks a lot.

By the way, the disk space issue arise from the fact of huge (not typical to any previous run) size of the current run folder not because I have low disk space.

Anyway I will follow suggested changes ones the current run get finished.

**Comment by maliberty** [comment]
I sorry I haven't had time to help with this more.  I think we need an examination of why this seems to require such heroics to get this block through.  @vijayank88 would you summarize the issues that have needed workarounds

**Comment by vijayank88** [comment]
@maliberty 
This design got successfully completed with our end and at user end its failed with 0.1 wns of hold violation.
While user trying to optimize and failed to complete the flow.
So its kept open to close by user and  they are trying to get final gds without any violations.
There is nothing pending on our side.

**Comment by maliberty** [comment]
@vijayank88 I'm looking at what we can learn to make this easier.   This process has taken over a month to complete and appears that the final result involves a fair amount of manual knob setting.

**Comment by vijayank88** [comment]
@Ryabkomv 
Any updates on timing violations?

**Comment by Ryabkomv** [comment]
@vijayank88 
with the above mentioned suggestion
```
set ::env(PL_RESIZER_HOLD_SLACK_MARGIN) {0.1}
set ::env(GLB_RESIZER_HOLD_SLACK_MARGIN) {0.1}
```

flow failed at the exact same step. 
I think it is not the scope of this issue, and more painful work is needed on violations investigation.

But I still wonder, why the flow generates enormous size of the `result ` folder (>300GB ) when I do not use `absolute ` sizing and as my disk size is 500 GB it finally fails, compared to several GB with suggested by you manual hack of fixed sizes of the chip and core. I'm not VLSI expert, is it usual way to provide the tool the size of the chip?

**Comment by maliberty** [comment]
@vijayank88 are you able to reproduce this large result?

**Comment by vijayank88** [comment]
@maliberty 
This my results directory disk usgare: `35GB`
```
149M	results/synthesis
169M	results/floorplan
477M	results/placement
750M	results/cts
6.1G	results/routing
0	results/eco
15G	results/signoff
188K	results/final/lef
3.0G	results/final/mag
136K	results/final/maglef
1.6G	results/final/def
3.1G	results/final/gds
790M	results/final/verilog/gl
790M	results/final/verilog
425M	results/final/spi/lvs
425M	results/final/spi
1.3G	results/final/spef
443M	results/final/sdf
140K	results/final/sdc
11G	results/final
35G	results
```

**Comment by Ryabkomv** [comment]
@vijayank88 I should re-run the flow as I have deleted the folder.
Have you completed the flow successfully without absolute sizing?

**Comment by maliberty** [comment]
35Gb vs 500 is quite a difference.   Can you two sort of what is going on?

**Comment by Ryabkomv** [comment]
@maliberty I started the flow and will update with the folder sizes and `config.tcl` once it fails

**Comment by Ryabkomv** [comment]
115G	results/floorplan
4 ,0K	results/cts
4 ,0K	results/routing
4 ,0K	results/signoff
4 ,0K	results/placement
4 ,0K	results/eco
115G	results/



158G    RUN_2022.08.01_10.31.29/issue_reproducible/openlane/designs/ExampleRocketSystem/runs
158G    RUN_2022.08.01_10.31.29/issue_reproducible/openlane/designs/ExampleRocketSystem
158G    RUN_2022.08.01_10.31.29/issue_reproducible/openlane/designs
158G    RUN_2022.08.01_10.31.29/issue_reproducible/openlane
13M     RUN_2022.08.01_10.31.29/issue_reproducible/pdk/sky130A/libs.ref/sky130_fd_sc_hd/lib
13M     RUN_2022.08.01_10.31.29/issue_reproducible/pdk/sky130A/libs.ref/sky130_fd_sc_hd
13M     RUN_2022.08.01_10.31.29/issue_reproducible/pdk/sky130A/libs.ref
13M     RUN_2022.08.01_10.31.29/issue_reproducible/pdk/sky130A
13M     RUN_2022.08.01_10.31.29/issue_reproducible/pdk
158G    RUN_2022.08.01_10.31.29/issue_reproducible
273G    RUN_2022.08.01_10.31.29/

config.tcl inside runs/tag/
```
set ::env(PDK_ROOT) {/home/ryabko/OpenLane/pdks}
set ::env(BASE_SDC_FILE) {/openlane/scripts/base.sdc}
set ::env(BOTTOM_MARGIN_MULT) {4}
set ::env(CARRY_SELECT_ADDER_MAP) {/home/ryabko/OpenLane/pdks/sky130A/libs.tech/openlane/sky130_fd_sc_hd/csa_map.v}
set ::env(CELLS_LEF) {/home/ryabko/OpenLane/pdks/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_fd_sc_hd.lef}
set ::env(CELLS_LEF_OPT) {/home/ryabko/OpenLane/pdks/sky130A/libs.ref/sky130_fd_sc_hd/lef/sky130_fd_sc_hd.lef}
set ::env(CELL_CLK_PORT) {CLK}
set ::env(CELL_PAD) {1}
set ::env(CELL_PAD_EXCLUDE) {sky130_fd_sc_hd__tap* sky130_fd_sc_hd__decap* sky130_ef_sc_hd__decap* sky130_fd_sc_hd__fill*}
set ::env(CHECK_ASSIGN_STATEMENTS) {0}
set ::env(CHECK_UNMAPPED_CELLS) {1}
set ::env(CLK_BUFFER) {sky130_fd_sc_hd__clkbuf_4}
set ::env(CLK_BUFFER_INPUT) {A}
set ::env(CLK_BUFFER_OUTPUT) {X}
set ::env(CLOCK_BUFFER_FANOUT) {16}
set ::env(CLOCK_PERIOD) {30.0}
set ::env(CLOCK_PORT) {clock}
set ::env(CLOCK_TREE_SYNTH) {1}
set ::env(CLOCK_WIRE_RC_LAYER) {met5}
set ::env(CONFIGS) {/openlane/configuration/extraction.tcl /openlane/configuration/synthesis.tcl /openlane/configuration/routing.tcl /openlane/configuration/general.tcl /openlane/configuration/cts.tcl /openlane/configuration/checkers.tcl /openlane/configuration/placement.tcl /openlane/configuration/lvs.tcl /openlane/configuration/floorplan.tcl}
set ::env(CTS_CLK_BUFFER_LIST) {sky130_fd_sc_hd__clkbuf_8 sky130_fd_sc_hd__clkbuf_4 sky130_fd_sc_hd__clkbuf_2}
set ::env(CTS_CLK_MAX_WIRE_LENGTH) {0}
set ::env(CTS_DISABLE_POST_PROCESSING) {0}
set ::env(CTS_DISTANCE_BETWEEN_BUFFERS) {0}
set ::env(CTS_MAX_CAP) {1.53169}
set ::env(CTS_REPORT_TIMING) {1}
set ::env(CTS_ROOT_BUFFER) {sky130_fd_sc_hd__clkbuf_16}
set ::env(CTS_SINK_CLUSTERING_MAX_DIAMETER) {50}
set ::env(CTS_SINK_CLUSTERING_SIZE) {25}
set ::env(CTS_SQR_CAP) {0.258e-3}
set ::env(CTS_SQR_RES) {0.125}
set ::env(CTS_TARGET_SKEW) {200}
set ::env(CTS_TECH_DIR) {N/A}
set ::env(CTS_TOLERANCE) {100}
set ::env(DATA_WIRE_RC_LAYER) {met2}
set ::env(DECAP_CELL) {sky130_ef_sc_hd__decap_12 sky130_fd_sc_hd__decap_8 sky130_fd_sc_hd__decap_6 sky130_fd_sc_hd__decap_4 sky130_fd_sc_hd__decap_3}
set ::env(DEFAULT_MAX_TRAN) {0.75}
set ::env(DEF_UNITS_PER_MICRON) {1000}
set ::env(DESIGN_CONFIG) {/openlane/designs/ExampleRocketSystem/config.tcl}
set ::env(DESIGN_DIR) {/openlane/designs/ExampleRocketSystem}
set ::env(DESIGN_IS_CORE) {1}
set ::env(DESIGN_NAME) {ExampleRocketSystem}
set ::env(DETAILED_ROUTER) {tritonroute}
set ::env(DIODE_CELL) {sky130_fd_sc_hd__diode_2}
set ::env(DIODE_CELL_PIN) {DIODE}
set ::env(DIODE_INSERTION_STRATEGY) {4}
set ::env(DIODE_PADDING) {2}
set ::env(DRC_EXCLUDE_CELL_LIST) {/home/ryabko/OpenLane/pdks/sky130A/libs.tech/openlane/sky130_fd_sc_hd/drc_exclude.cells}
```


config.tcl 


```
set ::env(DESIGN_NAME) ExampleRocketSystem
#set ::env(FP_SIZING) "absolute"
#set ::env(DIE_AREA) {0.0 0.0 6937.49 6948.21}
#set ::env(CORE_AREA) {5.52 10.88 6931.7 6935.04}
set ::env(FP_CORE_UTIL) 0.01

set ::env(CELL_PAD) 1

set ::env(PL_BASIC_PLACEMENT) 0
set ::env(PL_TARGET_DENSITY) "0.22"

set ::env(GLB_RT_ADJUSTMENT) 0.15

#set ::env(PL_ROUTABILITY_DRIVEN) 1
#set ::env(SYNTH_STRATEGY) "AREA 0"
# Change if needed
set ::env(VERILOG_FILES) [glob $::env(DESIGN_DIR)/src/*.v]

# Fill this
set ::env(CLOCK_PERIOD) "30.0"
set ::env(CLOCK_PORT) "clock"
#set ::env(CLOCK_NET) "clock"
set ::env(DIODE_INSERTION_STRATEGY) {4}
#set ::env(GLB_RT_MAXLAYER)          5
#set ::env(GLB_RESIZER_HOLD_SLACK_MARGIN) 0.2

#set ::env(GLB_RT_ADJUSTMENT) 0.25
#set ::env(PL_RESIZER_TIMING_OPTIMIZATIONS) 1
#set ::env(PL_RESIZER_HOLD_SLACK_MARGIN) {0.1}
#set ::env(GBL_RESIZER_HOLD_SLACK_MARGIN) {0.1}
#set ::env(PL_RESIZER_ALLOW_SETUP_VIOS) 1
set ::env(SYNTH_MAX_FANOUT) 10
set ::env(QUIT_ON_MAGIC_DRC) 0
set filename $::env(DESIGN_DIR)/$::env(PDK)_$::env(STD_CELL_LIBRARY)_config.tcl
if { [file exists $filename] == 1} {
        source $filename
}

```

**Comment by vijayank88** [comment]
@Ryabkomv 
`make survey || python3 ./env.py issue-survey` can post log of this

**Comment by Ryabkomv** [comment]
```
python3 ./env.py issue-survey
Kernel: Linux v5.4.0-122-generic
Distribution: ubuntu 20.04
Python: v3.8.10 (OK)
Container Engine: docker v20.10.17 (OK)
OpenLane Git Version: b718fd064063333703734ce3ae8bb56aa6810b0b
pip: INSTALLED
pip:venv: INSTALLED
---
PDK Version Verification Status: OK
---
Git Log (Last 3 Commits)

b718fd0 2022-07-05T10:05:16+03:00 Remove stale comment about power pin and net being the same (#1184) - Anton Blanchard -  (HEAD -> master, tag: 2022.07.06_01.50.05, origin/master, origin/HEAD)
b63f46b 2022-07-05T08:35:35+03:00 Added missing `-log` to `scrot_klayout` call (#1182) - Syndace -  ()
f9b5781 2022-07-01T16:04:31+02:00 Fix a bug with `-overwrite` (#1171) - Anton Blanchard -  (tag: 2022.07.02_01.38.08)

```

**Comment by vijayank88** [comment]
> @Ryabkomv Goto `OpenLane` folder and do following to commit latest updates from OpenLane.
> 
> ```
> git checkout master
> git pull
> make
> ```
> 
> If you're running out of memory remove previous `runs` from your directory. As per latest OpenLane use following `config.tcl` to run rocket chip
> 
> ```
> # User config
> set ::env(DESIGN_NAME) ExampleRocketSystem
> 
> # Change if needed
> set ::env(VERILOG_FILES) [glob $::env(DESIGN_DIR)/src/*.v]
> 
> # Fill this
> set ::env(CLOCK_PERIOD) "30.0"
> set ::env(CLOCK_PORT) "clock"
> set ::env(SYNTH_MAX_FANOUT) 10
> 
> set ::env(FP_CORE_UTIL) 0.01
> set ::env(PL_TARGET_DENSITY) 0.22
> set ::env(DIODE_INSERTION_STRATEGY) {4}
> set ::env(DPL_CELL_PADDING) 1
> set ::env(GRT_ADJUSTMENT) 0.15
> ```

@Ryabkomv 
I'm trying to help you, but you missing important updates and follow-up message from my side.

Please update OpenLane commits to latest and run the flow again with changes above mentioned configuration.

There was bug in the OpenLane commit you're using `2022-07-05` and its causing huge space utilized for `issue_reproducible`.

Its difficult for all to understand what's current issue. So i'm closing this issue and open new issue and there is any issue wrt flow completion.

Thanks.

**Comment by Ryabkomv** [comment]
I'm sorry I was missing to update the OpenLane.
I have updated and changed config.tcl.

**Comment by Ryabkomv** [comment]
Ok, same crush due to enormous disk space usage!

**Comment by donn** [comment]
please open a new issue

**Comment by Ryabkomv** [comment]
Guys, I don't want to be impolite, English is not my first language, so if you think my messagу has an intention to insult somebody, this is not correct, and may seem due to my immature english. You did a great product!

I tried to get the result for almost 2 months! I applied all suggested hacks in files somewhere deep inside installation. And still I had no one successful completion of the flow.

So I want to ask the question one more time: is there any example of successful application of OpenLane for real world processor? I try to generate the simplest Rocket core even not thinking about multy core BOOM processor. I understand I'm not using OpenRam or DFFRAM which can save area and load on the flow [several times](https://github.com/AUCOHL/DFFRAM), but again, I'm using simplest single Rocket core with 1 kB cash.

Maybe OpenLane is not mature for huge designs and it works well for processor sizes like pico32 provided as an example? This is perfectly fine, but I will not try to fight with impossible, saving lot of my and yours time. 
Or I just so unlucky...

**Comment by vijayank88** [comment]
@Ryabkomv 
Can you share gmail ID? 
I will  share my final results, what I generated with OpenLane flow.
We already had one teams meeting to help you on fixing `config.tcl` issue. But what is happening remotely at your end, we can't understand.

All I can say is with OpenLane tag [2022.07.30](https://github.com/The-OpenROAD-Project/OpenLane/releases/tag/2022.07.30) I'm able to complete routing and signoff check still running(More than 48 Hrs)

**Comment by Ryabkomv** [comment]
my gmail ID is RyabkoMV

**Comment by vijayank88** [comment]
Here is final GDS file:
![Screenshot from 2022-08-03 13-35-59](https://user-images.githubusercontent.com/85787913/182557273-d52ece37-fbad-45b1-bc18-7545047b6f46.png)

**Comment by maliberty** [comment]
If it is help we do have a nightly test design that is rocket based, see

https://jenkins.openroad.tools/job/OpenROAD-flow-scripts-Nightly-Public/1605/execution/node/656/log/?consoleFull

It is run on nangate45 rather than skywater.  We also run many other riscv cores like ibex, swerv-wrapper, and black parrot.

**Comment by Ryabkomv** [comment]
@maliberty 
what you provide with a link is open road flow script.
As I have understood from this issue, this is different from OpenLane, as the design succeeded with open road failed with OpenLane.

search for:
I tried with 30ns = 33.33MHz. Its passing with ORFS flow.
OpenLane Its crashed during global routing. Need to debug that.

in this issue.

**Comment by QuantamHD** [comment]
@Ryabkomv can you show me the contents of the floorplanning folder and its file sizes? Nothing in the floorplanning stage should be that large.

**Comment by maliberty** [comment]
I'm going to reopen this for the purposes of digging a bit deeper into the root causes of why this design is giving so much trouble closing.

I have been tied up with other issues so I'm just getting time now to jump in.  I'm sorry I couldn't help earlier in the process.

**Comment by Ryabkomv** [comment]
@QuantamHD 
```
ryabko@nanophotonics:~/OpenLane$ du -h designs/ExampleRocketSystem/runs/RUN_2022.08.02_07.42.11/results/floorplan/
115G	designs/ExampleRocketSystem/runs/RUN_2022.08.02_07.42.11/results/floorplan/
ryabko@nanophotonics:~/OpenLane$ ls -l designs/ExampleRocketSystem/runs/RUN_2022.08.02_07.42.11/results/floorplan/
total 119781104
-rw-r--r-- 1 root root 122536061461 авг  2 17:09 ExampleRocketSystem.def
```

**Comment by QuantamHD** [comment]
Wow that's a very large DEF... I don't think I've actually seen one this large to be honest.

@Ryabkomv could you use the less command to view that file. 

$ less  ExampleRocketSystem.def

Scroll through a bit and see if there's something of similar pattern that's consuming the space.

Could you also try to compress the file with tar or bzip2, and see if you can upload it Google drive?

@maliberty that seems like some kind of bug.

**Comment by QuantamHD** [comment]
Also do you know roughly how many standard cells are in the rocket core you generated?

**Comment by maliberty** [comment]
Your statement that you are using "1 kB cash" is not quite accurate.  
```
data_arrays_0_ext:  reg [31:0] ram [4095:0];
tag_array_ext: reg [20:0] ram [63:0];
data_arrays_0_0_ext: reg [31:0] ram [1023:0];
mem_ext: reg [31:0] ram [1023:0];
```
In your yosys log I see:
```
   Number of cells:             633680
     $_MUX_                     359438
     sky130_fd_sc_hd__dfxtp_2   175917
```
You have a huge amount of registers and muxes.  The version of tinyRocket we run in ORFS/nangate45 with RAMs has   less than 10k or each.  Admittedly it is not exactly the same netlist but it isn't this wildly different.

I think you have dismissed a real issue here.  
[Note that we have run multi-core BOOM in ORFS though it is not part of our nightly build due to size.]

**Comment by maliberty** [comment]
Total instances in ng45/tinyRocket is 25k

**Comment by maliberty** [comment]
In data_arrays_0_ext
```
  always @(posedge RW0_clk)
    if (RW0_en && RW0_wmode) begin
      for(i=0;i<4;i=i+1) begin
        if(RW0_wmask[i]) begin
          ram[RW0_addr][i*8 +: 8] <= RW0_wdata[i*8 +: 8];
        end
      end
    end
```
is going to generate a lot of muxes.  Similar code exists in data_arrays_0_0_ext.   These are behavioral models and not great for synthesizing.

**Comment by maliberty** [comment]
Anton was on the money with: 
![image](https://user-images.githubusercontent.com/761514/182689297-dc1f346d-ac33-4eb3-9e55-3bde517a74fe.png)

**Comment by Ryabkomv** [comment]
I admit, that using flip flops is much more resource consuming.

I have generated sample memory with OpenRam, so will try to use it some day
once understand how to do that.

If the synthesis of such design with flip flops is not possible, or not
supposed to be, or not guaranteed, just let me know.

**Comment by QuantamHD** [comment]
@maliberty I'm 100% for making this a stress test of OpenROAD. We should make it work even though DFFRAM might potentially be a way out.

**Comment by QuantamHD** [comment]
@maliberty Is 633680 the total number of mapped cells in the design? If so that seems on the smaller side of things.

**Comment by maliberty** [comment]
For comparison I gutted the 4 memory modules and then yosys shows:
```
   Number of cells:              61932
```
so virtual the whole design consists to the memories.

**Comment by maliberty** [comment]
I looked at the log files you uploaded a while back and the flow goes through P&R to the end.  I thought this was a failing run.  Do you have a current setup for this that shows the problem?

**Comment by maliberty** [comment]
I am running this as it to get a sense of the problem based on the config provided.  However I feel the real problem here is lowering the barrier to dropping in a RAM so that there is no incentive to go do such an unproductive path.  There is no way this can scale up to anything useful beyond the "1k cache".

@QuantamHD its not that large for a stress test and quite uninteresting as a design.  I would rather stress test with a more meaningful design than endless ff & muxes.

**Comment by QuantamHD** [comment]
Why is the DEF so large though? That makes no sense for this scale after floorplanning no less

**Comment by maliberty** [comment]
This is what I get in my run:
```
  -rw-r--r-- 1 mliberty mliberty 177147824 Aug  3 14:23 ExampleRocketSystem.def
```
How do you produce the DEF with that size?

**Comment by vijayank88** [comment]
@maliberty 

The design got completed in my gcp machine without changing RTL as @tspyrou  suggestion. Please let me know if you need any results or logs.
```
[INFO]: Calculating runtime...
[INFO]: Saving runtime environment...
[INFO]: Generating final set of reports...
[INFO]: Created manufacturability report at 'designs/ExampleRocketSystem/runs/Aug01/reports/manufacturability.rpt'.
[INFO]: Created metrics report at 'designs/ExampleRocketSystem/runs/Aug01/reports/metrics.csv'.
[WARNING]: There are max slew violations in the design at the typical corner. Please refer to 'designs/ExampleRocketSystem/runs/Aug01/reports/signoff/31-rcx_sta.slew.rpt'.
[WARNING]: There are max fanout violations in the design at the typical corner. Please refer to 'designs/ExampleRocketSystem/runs/Aug01/reports/signoff/31-rcx_sta.slew.rpt'.
[INFO]: There are no hold violations in the design at the typical corner.
[INFO]: There are no setup violations in the design at the typical corner.
[SUCCESS]: Flow complete.
[INFO]: Note that the following warnings have been generated:
[WARNING]: There are max slew violations in the design at the typical corner. Please refer to 'designs/ExampleRocketSystem/runs/Aug01/reports/signoff/31-rcx_sta.slew.rpt'.
[WARNING]: There are max fanout violations in the design at the typical corner. Please refer to 'designs/ExampleRocketSystem/runs/Aug01/reports/signoff/31-rcx_sta.slew.rpt'.
```

Total run time is: 58h47m23s0ms
Routing run time: 14h4m48s0ms
cell coutn: 751692

**Comment by Ryabkomv** [comment]
Let me summarize the history.
I started to dive into OpenLane with simple designs of ALU and small processors like pico32. It was successful!

Then I tried on Rocket core with tiny config. It failed at step 15 due to some reasons. I was suggested to modify config file, however still it failed at step 15 however with different errors.

Then I was suggested to modify https://github.com/The-OpenROAD-Project/OpenLane/blob/master/scripts/openroad/groute.tcl. Still the flow was not succesfull.

I had a teams call with vijayank88, and he found that pdk files overwrite config data. We just modified pdk *.tcl file.

Finally vijayank88 suggested config file, which allows me to get up to step 41 or so stopping due to hold violations.
This is not your job to resolve hold violations, but anyway, vijayank88 suggested some changes to config which doesn't help at my side.

**Important: last modifications used absolute sizing suggested by vijayank88**

```
set ::env(FP_SIZING) "absolute"
set ::env(DIE_AREA) {0.0 0.0 6937.49 6948.21}
set ::env(CORE_AREA) {5.52 10.88 6931.7 6935.04}
```
if I do not use absolute size, I got this enormous file size. The reason is not clear, but with my early attempts I do not use absolute sizing and I do not get such huge file size.

current config file:
```
# User config
set ::env(DESIGN_NAME) ExampleRocketSystem
#set ::env(FP_SIZING) "absolute"
#set ::env(DIE_AREA) {0.0 0.0 6937.49 6948.21}
#set ::env(CORE_AREA) {5.52 10.88 6931.7 6935.04}
set ::env(FP_CORE_UTIL) 0.01

#set ::env(CELL_PAD) 1

#set ::env(PL_BASIC_PLACEMENT) 0
set ::env(PL_TARGET_DENSITY) "0.22"

#set ::env(GLB_RT_ADJUSTMENT) 0.15

#set ::env(PL_ROUTABILITY_DRIVEN) 1
#set ::env(SYNTH_STRATEGY) "AREA 0"
# Change if needed
set ::env(VERILOG_FILES) [glob $::env(DESIGN_DIR)/src/*.v]

# Fill this
set ::env(CLOCK_PERIOD) "30.0"
set ::env(CLOCK_PORT) "clock"
#set ::env(CLOCK_NET) "clock"
set ::env(DIODE_INSERTION_STRATEGY) {4}
#set ::env(GLB_RT_MAXLAYER)          5
#set ::env(GLB_RESIZER_HOLD_SLACK_MARGIN) 0.2

#set ::env(GLB_RT_ADJUSTMENT) 0.25
#set ::env(PL_RESIZER_TIMING_OPTIMIZATIONS) 1
#set ::env(PL_RESIZER_HOLD_SLACK_MARGIN) {0.1}
#set ::env(GBL_RESIZER_HOLD_SLACK_MARGIN) {0.1}
#set ::env(PL_RESIZER_ALLOW_SETUP_VIOS) 1
set ::env(SYNTH_MAX_FANOUT) 10
#set ::env(QUIT_ON_MAGIC_DRC) 0
set ::env(DPL_CELL_PADDING) 1
set ::env(GRT_ADJUSTMENT) 0.15

set filename $::env(DESIGN_DIR)/$::env(PDK)_$::env(STD_CELL_LIBRARY)_config.tcl
if { [file exists $filename] == 1} {
        source $filename
}
```
this are first lines from ExampleRocketSystem.def

```
VERSION 5.8 ;
DIVIDERCHAR "/" ;
BUSBITCHARS "[]" ;
DESIGN ExampleRocketSystem ;
UNITS DISTANCE MICRONS 1000 ;
DIEAREA ( 0 0 ) ( 291667100 291677820 ) ;
ROW ROW_0 unithd 5520 10880 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_1 unithd 5520 13600 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_2 unithd 5520 16320 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_3 unithd 5520 19040 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_4 unithd 5520 21760 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_5 unithd 5520 24480 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_6 unithd 5520 27200 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_7 unithd 5520 29920 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_8 unithd 5520 32640 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_9 unithd 5520 35360 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_10 unithd 5520 38080 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_11 unithd 5520 40800 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_12 unithd 5520 43520 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_13 unithd 5520 46240 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_14 unithd 5520 48960 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_15 unithd 5520 51680 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_16 unithd 5520 54400 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_17 unithd 5520 57120 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_18 unithd 5520 59840 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_19 unithd 5520 62560 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_20 unithd 5520 65280 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_21 unithd 5520 68000 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_22 unithd 5520 70720 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_23 unithd 5520 73440 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_24 unithd 5520 76160 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_25 unithd 5520 78880 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_26 unithd 5520 81600 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_27 unithd 5520 84320 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_28 unithd 5520 87040 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_29 unithd 5520 89760 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_30 unithd 5520 92480 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_31 unithd 5520 95200 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_32 unithd 5520 97920 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_33 unithd 5520 100640 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_34 unithd 5520 103360 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_35 unithd 5520 106080 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_36 unithd 5520 108800 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_37 unithd 5520 111520 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_38 unithd 5520 114240 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_39 unithd 5520 116960 FS DO 634034 BY 1 STEP 460 0 ;
ROW ROW_40 unithd 5520 119680 N DO 634034 BY 1 STEP 460 0 ;
ROW ROW_41 unithd 5520 122400 FS DO 634034 BY 1 STEP 460 0 ;
:

```

**Comment by Ryabkomv** [comment]
Also I updated [log files](https://github.com/Ryabkomv/openroad_issue_reproducible/blob/master/logs.tar.gz) from the last run

**Comment by maliberty** [comment]
Thanks for the history.  I feel that too much time is being spent on pursuing a path that is known to give a suboptimal result even if you manage to get over all the hurdles.  Dropping in a ram (dff or openram) would likely solve many of your issues and give a better result, likely with less effort.

**Comment by vijayank88** [comment]
@Ryabkomv 
Any progress on this at your end?

**Comment by Ryabkomv** [comment]
If I understand last message right way, I was suggested not to use open
lane for such big number of gates. Is anything expected from my side in
this regard?

**Comment by maliberty** [comment]
As Anton and I both suggested earlier you should switch to real rams for your memories.  This block is unnecessarily dense due to your decision not to.   

OL should also drop checking max fanout at the end of the flow as I've also suggested.

The need for the alpha=0 workaround will eventually be addressed in OR but isn't a blocking issue.
